transports/tcp/src/lib.rs

```diff
@@ -734,116 +710,84 @@ where
-                            match ev {
-                                Ok(IfEvent::Up(inet)) => {
-                                    let ip = inet.addr();
-                                    if me.listen_addr.is_ipv4() == ip.is_ipv4() && addrs.insert(ip)
-                                    {
-                                        let ma = ip_to_multiaddr(ip, me.listen_addr.port());
-                                        log::debug!("New listen address: {}", ma);
-                                        me.port_reuse.register(ip, me.listen_addr.port());
-                                        return Poll::Ready(Some(Ok(
-                                            TcpListenerEvent::NewAddress(ma),
-                                        )));
-                                    }
-                                }
-                                Ok(IfEvent::Down(inet)) => {
-                                    let ip = inet.addr();
-                                    if me.listen_addr.is_ipv4() == ip.is_ipv4() && addrs.remove(&ip)
-                                    {
-                                        let ma = ip_to_multiaddr(ip, me.listen_addr.port());
-                                        log::debug!("Expired listen address: {}", ma);
-                                        me.port_reuse.unregister(ip, me.listen_addr.port());
-                                        return Poll::Ready(Some(Ok(
-                                            TcpListenerEvent::AddressExpired(ma),
-                                        )));
-                                    }
-                                }
-                                Err(err) => {
-                                    log::debug! {
-                                        "Failure polling interfaces: {:?}. Scheduling retry.",
-                                        err
-                                    };
-                                    me.pause = Some(Delay::new(me.sleep_on_error));
-                                    return Poll::Ready(Some(Ok(TcpListenerEvent::Error(err))));
-                                }
-                            }
-                        }
-                    }
-                },
-                // If the listener is bound to a single interface, make sure the
-                // address is registered for port reuse and reported once.
-                InAddr::One { addr, out } => {
-                    if let Some(multiaddr) = out.take() {
-                        me.port_reuse.register(*addr, me.listen_addr.port());
-                        return Poll::Ready(Some(Ok(TcpListenerEvent::NewAddress(multiaddr))));
-                    }
+        if let Some(mut pause) = me.pause.take() {
+            match pause.poll_unpin(cx) {
+                Poll::Ready(_) => {}
+                Poll::Pending => {
+                    me.pause = Some(pause);
+                    return Poll::Pending;
```

Comment:
I moved this section to the top so that the pause is polled before polling the `if_watcher`. 
If polling the `if_watcher` errored, we set a delay before returning. From the logged error messages _"Failure polling interfaces: {..}. Scheduling retry."_ I assume that the if_watcher should be polled again **after** this delay. Thus on the next iteration we should first poll the pause. Is this assumption correct?

Reply:
Seems reasonable to me but then we should maybe put the loop around everything and poll this more often?

Reply:
@mxinden 
> Does anyone have further insight? If not, I am happy to look deeper into this.

I was wondering about that as well, but I also don't have any further insights.

@thomaseizinger 
> Seems reasonable to me but then we should maybe put the loop around everything and poll this more often?

Each time the delay is set we directly `return` anyway, so there is no case where we would `continue` the loop and `pause` is `Some`.

Reply:
Let's move the discussion on whether the delay is needed in the first place to https://github.com/libp2p/rust-libp2p/pull/2849.